### PR TITLE
WIP: update-extra-deps

### PR DESCRIPTION
**Pull request by zudov**
Already works, but dirty and slow #1

### CODE REVIEW

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+data ExtraDep = ExtraDep
+  { _packageName :: Text
+  , _version     :: Version
+  } deriving (Show, Eq)
+
+pPrintExtraDep :: ExtraDep -> Text
+pPrintExtraDep (ExtraDep packageName version) =
+  packageName <> "-" <> pPrintVersion version
+
+newtype Version = Version { unVersion :: [Int] } deriving (Show, Eq, Ord)
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Why not `Version` from `Data.Version`/`Distribution.Version` ?

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        word = concat <$> sequence [ many Atto.digit
+                                   , pure <$> Atto.letter
+                                   , many (Atto.letter <|> Atto.digit)
+                                   ]
+    pVersion = Version <$> (((:)) <$> num <*> many num')
+      where
+        num = Atto.decimal
+        num' = Atto.char '.' *> num
+
+data ExtraDep = ExtraDep
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: http://hackage.haskell.org/package/Cabal-1.22.7.0/docs/Distribution-Package.html#t:PackageIdentifier

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    traverseOf (key "extra-deps" . values . _String) updateExtraDep
+  where
+    updateExtraDep :: Text -> IO Text
+    updateExtraDep dep =
+      either
+        (\err -> dep <$ putStrLn err)
+        (fmap pPrintExtraDep . getNewest)
+        (Atto.parseOnly (pExtraDep <* Atto.endOfInput) dep)
+
+pExtraDep :: Atto.Parser ExtraDep
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Actually, should we just `regex-applicative-text` here. Would be much clearer then attoparsec (and can be more resilient as well)

_EDIT_ seems I use it already anyway :)

Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Yes, `regex-applicative` would match it better. I'll switch to it now.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 1 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Yes, need to switch to those existing datatypes.

File: package.yaml [hunk 4]
```diff
@@ -30,8 +30,14 @@ dependencies:
 - regex-applicative-text
 - vector
 - yaml
+- attoparsec
```
Comment (phadej) [inline package.yaml]: sort the deps (case-insensitive when ready)

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    traverseOf
+      (key "extra-deps" . values . _String)
+      (updateExtraDep newest)
+      value
+
+  where
+    updateExtraDep :: PackDeps.Newest -> Text -> IO Text
+    updateExtraDep newest dep =
+      either
+        (\err -> dep <$ putStrLn err)
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: hPutStrLn stderr

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        num' = Atto.char '.' *> num
+
+data ExtraDep = ExtraDep
+  { _packageName :: Text
+  , _version     :: Version
+  } deriving (Show, Eq)
+
+pPrintExtraDep :: ExtraDep -> Text
+pPrintExtraDep (ExtraDep packageName version) =
+  packageName <> "-" <> pPrintVersion version
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: 4 space indent

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 1 above)
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: :+1:

Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 2 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 6 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed (in some other places).

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 5 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed

**Comment by zudov** [comment]
Before using that command user would need to do `cabal update`. If they don't it would result into failures like:

```
*** Exception: /home/zudov/.cabal/config: openBinaryFile: does not exist (No such file or directory)
```

or

```
*** Exception: /home/zudov/.cabal/packages/hackage.haskell.org/00-index.tar: openBinaryFile: does not exist (No such file or directory)
```

I think these are familiar for most haskellers hints to run `cabal update`, so we probably can just leave it like that.

We probably could mention it in the README or command description.

**Comment by phadej** [comment]
Could you try to `stack install packdeps` and check whether it catches the error. Maybe we could catch the error, and abort a bit more gracefully.

**Comment by zudov** [comment]
It's the same with `packdeps`:

``` haskell
$ packdeps stack-yaml.cabal 
packdeps: /home/zudov/.cabal/config: openFile: does not exist (No such file or directory)
$ stack exec stack-yaml -- update-extra-deps
stack-yaml: /home/zudov/.cabal/config: openFile: does not exist (No such file or directory)
```

**Comment by zudov** [comment]
I'll catch the exception and add some helpful error message to it.

**Comment by zudov** [comment]
See https://github.com/zudov/stack-yaml/commit/c0efa041da53df5a8cdfeaf2194ed3600dcb806f

File: package.yaml (hunk 4 above)
Comment (zudov) [inline package.yaml]: Fixed

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+               (\name ->
+                   newestPackageId newest name
+                     & maybe
+                         (dep <$ hPutStrLn
+                                   stderr
+                                   ("Ignoring extra-deps entry "
+                                       <> show dep
+                                       <> " as it refers to unknown package "
+                                       <> show (unPackageName name)))
+                         (pure . pPrintPackageIdentifier))
```
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs:57]: I probably should just use `case` here, to be consistent with [`updateGithubDep`](https://github.com/phadej/stack-yaml/pull/2/files#diff-9cddc95048f32deafa05519a80386aa5R21)

Or we could rewrite `updateGithubDep` into something like:

``` haskell
updateGithubDep :: Object -> IO Object
updateGithubDep obj =
    ((obj ^? ix "git" . _String) >>= parseGithub)
      & maybe
          (pure obj)
          (\(owner, repo) ->
              (GH.executeRequest' $ GH.branchesForR owner repo $ Just 1)
                 & either
                     throwM
                     (\branches ->
                        (V.find ((== "master") . GH.branchName) branches)
                           & maybe
                               (pure obj)
                               (\branch ->
                                   let obj' = obj & ix "commit" . _String .~ GH.branchCommitSha)
                                   in pure obj'))
```

Which one do you think is better?

Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs:57]: Leave it as is. I'll clean-up towers-of-indentation later.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+parsePackageIdentifier = match pkgIdentifierRe
+  where
+    pkgIdentifierRe =
+      PackageIdentifier
+        <$> (PackageName <$> few anySym <* sym '-')
+        <*> pkgVersionRe
+
+    pkgVersionRe = flip Version [] <$> ((:) <$> num <*> many (sym '.' *> num))
+      where
+        num = read <$> many (psym isDigit)
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: http://hackage.haskell.org/package/regex-applicative-0.3.2.1/docs/Text-Regex-Applicative-Common.html#v:decimal

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                                       <> show dep
+                                       <> " as it refers to unknown package "
+                                       <> show (unPackageName name)))
+                         (pure . pPrintPackageIdentifier))
+
+parsePackageIdentifier :: Text -> Maybe PackageIdentifier
+parsePackageIdentifier = match pkgIdentifierRe
+  where
+    pkgIdentifierRe =
+      PackageIdentifier
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: I'd indent as

``` hs
    pkgIdentifierRe = PackageIdentifier
        <$> (PackageName <$> few anySym <* sym '-')
        <*> pkgVersionRe
```

File: src/StackYaml/Transformations/UpdateExtraDeps.hs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+    pkgVersionRe = flip Version [] <$> ((:) <$> num <*> many (sym '.' *> num))
+      where
+        num = read <$> many (psym isDigit)
+
+pPrintPackageIdentifier :: PackageIdentifier -> Text
+pPrintPackageIdentifier PackageIdentifier{..} =
+    pack (unPackageName pkgName) <> "-" <> pPrintVersion pkgVersion
+
+pPrintVersion :: Version -> Text
```
Comment (phadej) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: there is `showVersion` in `Data.Version` (probably re-exported from `Distribution.Version`

Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 8 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Huh. I was wondering where those things are and somehow didn't find it.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 10 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 8 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed.

File: src/StackYaml/Transformations/UpdateExtraDeps.hs (hunk 9 above)
Comment (zudov) [inline src/StackYaml/Transformations/UpdateExtraDeps.hs]: Fixed

**Comment by zudov** [comment]
Squashed

**Comment by phadej** [comment]
Thanks!

**Pull request merged**